app/services/activitypub/synchronize_followers_service.rb

```diff
@@ -50,21 +64,35 @@ def build_undo_follow_json(follow)
     Oj.dump(serialize_payload(follow, ActivityPub::UndoFollowSerializer))
   end
 
-  def collection_items(collection_or_uri)
+  # Yields to the block on each page
+  # Only returns true if the whole collection has been processed
+  def process_collection(collection_or_uri, max_pages: MAX_COLLECTION_PAGES)
     collection = fetch_collection(collection_or_uri)
-    return unless collection.is_a?(Hash)
+    return false unless collection.is_a?(Hash)
+
+    collection = collection['first'] if collection['first'].present?
+
+    while collection.is_a?(Hash) || collection.is_a?(String)
+      max_pages -= 1
+      return false if max_pages.negative?
 
-    collection = fetch_collection(collection['first']) if collection['first'].present?
-    return unless collection.is_a?(Hash)
+      collection = fetch_collection(collection)
+      return false unless collection.is_a?(Hash)
 
-    # Abort if we'd have to paginate through more than one page of followers
-    return if collection['next'].present?
+      yield as_array(collection_page_items(collection))
+
+      collection = collection['next'].presence
```

Comment:
I do not have any concrete idea of how this could be improved, but I found this very hard to follow, because `collection` sometimes is a Hash (presumably including the actual items) and sometimes a URI.

Splitting this into two distinct variables, one for the currently loaded collection(page) and one for the next page's URI might help explaining the flow here. But it might as well make this more complicated, so feel free to ignore this.

Reply:
It is! And you managed to make it even simpler. Amazing.